File: IT/1/Oggetti/Esercizi/Python/esercizio26.py
# file esercizio.py

from __future__ import print_function
from math import pow, sqrt

class Point2d(object):
	def __init__(self, x, y):
		self.__x = x
		self.__y = y

	def getX(self):
		return self.__x

	def setX(self, x):
		self.__x = x

	def getY(self):
		return self.__y

	def setY(self, y):
		self.__y = y

	def __str__(self):
		return "({}, {})".format(self.__x, self.__y)

	x = property(getX, setX)
	y = property(getY, setY)

class Point3d(Point2d):
	def __init__(self, x, y, z):
		super(Point3d, self).__init__(x, y)
		self.__z = z

	def getX(self):
		return self.__x

	def setZ(self, z):
		self.__z = z

	def getZ(self):
		return self.__z

	def __str__(self):
		return "({}, {}, {})".format(self.x, self.y, self.__z)

	z = property(getZ, setZ)

class Plane(object):
	def __init__(self, p1, p2, p3):
		self.__p1 = p1
		self.__p2 = p2
		self.__p3 = p3

	def __str__(self):
		return "Piano definito dai punti P1={}, P2={}, P3={}".format(self.__p1, self.__p2,self.__p3)

p1 = Point3d(0.0, 0.0, 0.0)
p2 = Point3d(1.0, 2.0, 3.0)
p3 = Point3d(3.0, 4.0, 5.0)

pi = Plane(p1=p1, p2=p2, p3=p3)

print(pi)

File: IT/1/Oggetti/Esercizi/C++/esercizio26.cpp
/* file esercizio26.cpp */

#include <iostream>
#include <cmath>

using std::cin;
using std::cout;
using std::endl;
using std::ostream;

class Point2d {
	private:
		double x;
		double y;

	public:
		Point2d(double, double);
		double getX() const;
		void setX(double);
		double getY() const;
		void setY(double);

	friend ostream& operator<<(ostream& os, const Point2d&);
};

Point2d::Point2d(double x, double y) {
	this->setX(x);
	this->setY(y);
}

double Point2d::getX() const {
	return this->x;
}

void Point2d::setX(double x) {
	this->x = x;
}

double Point2d::getY() const {
	return this->y;
}

void Point2d::setY(double y) {
	this->y = y;
}

ostream& operator<<(ostream& os, const Point2d& p) {
	return os << "(" << p.x << ", " << p.y << ")";
}

class Point3d : public Point2d {
	private:
		double z;

	public:
		Point3d(double, double, double);
		double getZ();
		void setZ(double);

	friend ostream& operator<<(ostream& os, const Point3d&);
};

Point3d::Point3d(double x, double y, double z) : Point2d(x, y) {
	this->setZ(z);
}

double Point3d::getZ() {
	return this->z;
}

void Point3d::setZ(double z) {
	this->z = z;
}

ostream& operator<<(ostream& os, const Point3d& p) {
	return os << "(" << p.getX() << ", " << p.getY() << ", " << p.z << ")";
}

class Plane {
	private:
		Point3d *p1;
		Point3d *p2;
		Point3d *p3;

	public:
		Plane(Point3d*, Point3d*, Point3d*);

	friend ostream& operator<<(ostream& os, const Plane&);
};

Plane::Plane(Point3d* p1, Point3d* p2, Point3d* p3) {
	this->p1 = p1;
	this->p2 = p2;
	this->p3 = p3;
}

ostream& operator<<(ostream& os, const Plane& pi) {
	return os << "Piano definito dai punti P1=" << *pi.p1 << ", P2=" << *pi.p2 << ", P3=" << *pi.p3;
}

int main() {
	Point3d p1(0.0, 0.0, 0.0);
	Point3d p2(1.0, 2.0, 3.0);
	Point3d p3(3.0, 2.0, 1.0);

	Plane pi(&p1, &p2, &p3);

	cout << pi << endl;
}

